// vnt/lib/utils/config.js
/*
    This file is part of vnt.js.

    vnt.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    vnt.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with vnt.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/** @file config.js
 * @authors:
 *   Marek Kotewicz <marek@ethdev.com>
 * @date 2015
 */

/**
 * Utils
 *
 * @module utils
 */

/**
 * Utility functions
 *
 * @class [utils] config
 * @constructor
 */


/// required to define VNT_BIGNUMBER_ROUNDING_MODE
var BigNumber = require('bignumber.js');

var VNT_UNITS = [
    'wei',
    'Kwei',
    'Mwei',
    'Gwei',
    'microvnt',
    'micro',
    'millivnt',
    'milli',
    'vnt'
];

module.exports = {
    VNT_PADDING: 32,
    VNT_SIGNATURE_LENGTH: 4,
    VNT_UNITS: VNT_UNITS,
    VNT_BIGNUMBER_ROUNDING_MODE: { ROUNDING_MODE: BigNumber.ROUND_DOWN },
    VNT_POLLING_TIMEOUT: 1000/2,
    defaultBlock: 'latest',
    defaultAccount: undefined
};
